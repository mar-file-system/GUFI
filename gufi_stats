#!/usr/bin/env python2
# This file is part of GUFI, which is part of MarFS, which is released
# under the BSD license.


# Copyright (c) 2017, Los Alamos National Security (LANS), LLC
# All rights reserved.

# Redistribution and use in source and binary forms, with or without modification,
# are permitted provided that the following conditions are met:

# 1. Redistributions of source code must retain the above copyright notice, this
# list of conditions and the following disclaimer.

# 2. Redistributions in binary form must reproduce the above copyright notice,
# this list of conditions and the following disclaimer in the documentation and/or
# other materials provided with the distribution.

# 3. Neither the name of the copyright holder nor the names of its contributors
# may be used to endorse or promote products derived from this software without
# specific prior written permission.

# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
# IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
# OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

# -----
# NOTE:
# -----

# GUFI uses the C-Thread-Pool library.  The original version, written by
# Johan Hanssen Seferidis, is found at
# https://github.com/Pithikos/C-Thread-Pool/blob/master/LICENSE, and is
# released under the MIT License.  LANS, LLC added functionality to the
# original work.  The original work, plus LANS, LLC added functionality is
# found at https://github.com/jti-lanl/C-Thread-Pool, also under the MIT
# License.  The MIT License can be found at
# https://opensource.org/licenses/MIT.


# From Los Alamos National Security, LLC:
# LA-CC-15-039

# Copyright (c) 2017, Los Alamos National Security, LLC All rights reserved.
# Copyright 2017. Los Alamos National Security, LLC. This software was produced
# under U.S. Government contract DE-AC52-06NA25396 for Los Alamos National
# Laboratory (LANL), which is operated by Los Alamos National Security, LLC for
# the U.S. Department of Energy. The U.S. Government has rights to use,
# reproduce, and distribute this software.  NEITHER THE GOVERNMENT NOR LOS
# ALAMOS NATIONAL SECURITY, LLC MAKES ANY WARRANTY, EXPRESS OR IMPLIED, OR
# ASSUMES ANY LIABILITY FOR THE USE OF THIS SOFTWARE.  If software is
# modified to produce derivative works, such modified software should be
# clearly marked, so as not to confuse it with the version available from
# LANL.

# THIS SOFTWARE IS PROVIDED BY LOS ALAMOS NATIONAL SECURITY, LLC AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
# THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL LOS ALAMOS NATIONAL SECURITY, LLC OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
# OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
# IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
# OF SUCH DAMAGE.



import argparse
import itertools
import os
import subprocess
import sys

import query_builder as qb

def build(args):
    select = [args.select] if args.select else []

    where = []
    if args.uid:
        where += ['uid == "{}"'.format(args.uid)]

    return select, where

def space(args, order=''):
    select, where = build(args)
    entries       = qb.build_query(select + ['uid', 'TOTAL(size) AS space'],  ['entries'], where, ['uid'], ['space {}'.format(order)], args.num_results)
    intermediate  = None
    aggregate     = qb.build_query(select + ['uid', 'TOTAL(space) AS space'], ['entries'], where, ['uid'], ['space {}'.format(order)], args.num_results)
    return entries, intermediate, aggregate

def files(args, order=''):
    select, where = build(args)
    entries       = qb.build_query(select + ['uid', 'COUNT(*) AS files'],     ['entries'], where, ['uid'], ['files {}'.format(order)], args.num_results)
    intermediate  = None
    aggregate     = qb.build_query(select + ['uid', 'TOTAL(files) AS files'], ['entries'], where, ['uid'], ['files {}'.format(order)], args.num_results)
    return entries, intermediate, aggregate

def total_space(args, *ignored):
    _, where = build(args)
    entries      = qb.build_query(['TOTAL(size) AS size'], ['entries'], where, None, None, None, None)
    intermediate = qb.build_query(['TOTAL(size) AS size'], ['entries'], None,  None, None, None, None)
    return entries, intermediate, intermediate

def total_files(args, *ignored):
    _, where = build(args)
    entries      = qb.build_query(['COUNT(*)     AS count'], ['entries'], where, None, None, None, None)
    intermediate = qb.build_query(['TOTAL(count) AS count'], ['entries'], None,  None, None, None, None)
    return entries, intermediate, intermediate

def directory_count(args, order = ''):
    _, where = build(args)
    entries        = qb.build_query(['COUNT(*) AS count'], ['entries'], where, None, None, None, None)
    aggregate      = qb.build_query(['count'],             ['entries'], where, None, ['count {}'.format(order)], args.num_results, None)

    return entries, None, aggregate

def get_leaf_dirs(path):
    '''Return an array containing path names of leaf directories under path'''
    return [dirpath for dirpath, dirnames, filenames in os.walk(path) if not dirnames]

def leaf_count(args, order = ''):
    args.GUFI_tree = list(itertools.chain.from_iterable([get_leaf_dirs(path) for path in args.GUFI_tree]))
    return directory_count(args, order)

def total_in_leaf(args, *ignored):
    _, where = build(args)
    entries        = qb.build_query(['COUNT(*)     AS count'], ['entries'], where, None, None, None, None)
    intermediate   = qb.build_query(['TOTAL(count) AS count'], ['entries'], None,  None, None, None, None)
    aggregate      = qb.build_query(['count'],                 ['entries'], None,  None, None, None, None)

    args.GUFI_tree = list(itertools.chain.from_iterable([get_leaf_dirs(path) for path in args.GUFI_tree]))

    return entries, intermediate, aggregate

def median_in_leaf(args, *ignored):
    '''https://stackoverflow.com/a/15766121'''
    _, where = build(args)
    entries      = qb.build_query(['COUNT(*) AS count'], ['entries'], where, None, None, None, None)
    intermediate = qb.build_query(['count'],             ['entries'], None,  None, ['count'], '2 - (SELECT COUNT(*) FROM entries) % 2', ['OFFSET (SELECT (COUNT(*) - 1) / 2 FROM entries)'])
    aggregate    = qb.build_query(['AVG(count)'],        ['entries'], None,  None, None, None, None)

    args.GUFI_tree = list(itertools.chain.from_iterable([get_leaf_dirs(path) for path in args.GUFI_tree]))

    return entries, intermediate, aggregate

def get_depth(path, sep = '/'):
    ret = []
    path_len = len(path)
    sub = path[-1] != sep # subtract 1 if the provided path did not include a separator because the split string will start with a separator

    for dirpath, _, _ in os.walk(path):
        ret += [[path, len(dirpath[path_len:].split(sep)) - sub]]
    return ret

def depth(args, sep = '/'):
    ret = []
    for path in args.GUFI_tree:
        ret += get_depth(path);
    return ret

def leaf_depth(args, sep = '/'):
    ret = []
    for path in args.GUFI_tree:
        for leaf in get_leaf_dirs(path):
            ret += get_depth(leaf, sep)
    return ret

# mapping of valid non-bfq 'operation' argument values
NONQUERIES  = {'depth'      : depth,
               'leaf-depth' : leaf_depth}

# mapping of valid bfq 'operation' argument values to array generators
BFQ_QUERIES = {'space'           : space,
               'files'           : files,
               'total-space'     : total_space,
               'total-files'     : total_files,
               'directory-count' : directory_count,
               'leaf-count'      : leaf_count,
               'total-leaf'      : total_in_leaf,
               'median-leaf'     : median_in_leaf}

# allowed keywords for sorting output (if applicable)
ASCENDING  = 'ASC'
DESCENDING = 'DESC'
ORDER = {'ASC'   : ASCENDING,
         'DESC'  : DESCENDING,
         'least' : ASCENDING,
         'most'  : DESCENDING}

if __name__=='__main__':
    cwd = os.getcwd()

    parser = argparse.ArgumentParser(description='GUFI statistics')
    parser.add_argument('--order',         metavar='order', dest='order',       choices=ORDER.keys(),        default='',           help='sort output (if applicable)')
    parser.add_argument('--delim',         metavar='c',     dest='delim',       type=qb.check_char,                                help='delimiter separating output columns')
    parser.add_argument('--maxdepth',      metavar='d',     dest='maxdepth',    type=qb.check_non_negative,                        help='the maximum number of levels to search')
    parser.add_argument('--mindepth',      metavar='d',     dest='mindepth',    type=qb.check_non_negative,                        help='the minimum number of levels to search')
    parser.add_argument('--num_results',   metavar='n',     dest='num_results', type=qb.check_non_negative,                        help='first n results')
    parser.add_argument('--select',        metavar='cols',  dest='select',      type=str,                                          help='comma separated columns')
    parser.add_argument('--threads',       metavar='n',     dest='threads',     type=qb.check_non_negative,  default=qb.cpus(),    help='number of threads')
    parser.add_argument('--uid', '--user', metavar='u',     dest='uid',         type=qb.get_uid,                                   help='restrict to user')
    parser.add_argument('operation',                                            choices=NONQUERIES.keys() + BFQ_QUERIES.keys(),    help='operation to perform')
    parser.add_argument('GUFI_tree',                                                                         nargs='+',            help='path containing GUFI tree database file')

    args = parser.parse_args();

    if args.operation in NONQUERIES:
        res = NONQUERIES[args.operation](args)
        res.sort(key = lambda pair: pair[1], reverse=(args.order==DESCENDING))
        print len(res[:args.num_results])
    elif args.operation in BFQ_QUERIES:
        entries, intermediate, aggregate = BFQ_QUERIES[args.operation](args, ORDER[args.order] if args.order in ORDER else '')

        # create the bfq command
        bfq_cmd = ['{}/bfq'.format(os.path.dirname(os.path.realpath(__file__))),
                   '-n', str(args.threads),
                   '-p', '-P']

        if entries is not None:
            bfq_cmd += ['-E', entries]

        if intermediate is not None:
            bfq_cmd += ['-J', intermediate]

        if aggregate is not None:
            bfq_cmd += ['-G', aggregate]

        if args.delim:
            bfq_cmd += ['-d', args.delim]

        if args.maxdepth:
            bfq_cmd += ['-z', str(args.maxdepth)]

        if args.mindepth:
            bfq_cmd += ['-y', str(args.mindepth)]

        bfq = subprocess.Popen(bfq_cmd + args.GUFI_tree)
        bfq.communicate() # block until bfq finishes

        sys.exit(bfq.returncode)
