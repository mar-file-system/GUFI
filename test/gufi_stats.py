#!/usr/bin/env python2

import imp
import os
import sys

file_dir = os.path.dirname(os.path.abspath(__file__))
root_dir = os.path.dirname(file_dir)
gufi_stats = imp.load_source('gufi_stats', os.path.join(root_dir, 'scripts', 'gufi_stats'))

if __name__=='__main__':
    sys.exit(gufi_stats.run(sys.argv, "config.test"))
