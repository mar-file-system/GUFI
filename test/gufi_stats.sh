#!/usr/bin/env bash

set -e

ROOT="$(realpath $(dirname ${BASH_SOURCE[0]})/..)"
export PATH=${ROOT}/scripts:${PATH}
export PYTHONPATH=${ROOT}/scripts:${PYTHONPATH}

GUFI_STATS="${ROOT}/test/gufi_stats.py"

function get_next() {
    awk -F: "{id[\$3]=1}END{i = $1; while (i < 1000000) {if(id[i] == \"\"){print i; exit;} i++}}" "$2"
}

# find 2 unused UIDs
# UID1 should be the caller, since a non-existant UID would not be able to access this directory
UID1=$(id -u)
UID2=$(get_next $(( ${UID1} > 1004?(${UID1} + 1):1004 )) /etc/passwd)
UID3=$(get_next $((${UID2} + 1))                         /etc/passwd)

# find 2 unused IDs
GID1=$(id -g)
GID2=$(get_next $(( ${GID1} > 1004?(${GID1} + 1):1004 )) /etc/group)
GID3=$(get_next $((${GID2}  + 1))                        /etc/group)

CONFIG="config.test"
OUTPUT="gufi_stats.out"

function cleanup {
    sudo rm -rf "${CONFIG}" "${OUTPUT}" prefix prefix.gufi
}

trap cleanup EXIT

# generate a fake config file
echo "Threads=$(nproc)" > ${CONFIG}
echo "Exec=${ROOT}/src/gufi_query" >> ${CONFIG}
echo "IndexRoot=$(pwd)/prefix.gufi" >> ${CONFIG}
sudo chmod 666 ${CONFIG}

# generate the tree
sudo rm -rf prefix prefix.gufi
mkdir prefix
mkdir -p prefix/${UID1}
mkdir -p prefix/${UID2}
mkdir -p prefix/${UID3}
mkdir prefix/${UID2}/dir
truncate -s 1 prefix/${UID1}/f1
truncate -s 2 prefix/${UID2}/f1
truncate -s 2 prefix/${UID2}/f2
truncate -s 2 prefix/${UID2}/dir/f1
truncate -s 2 prefix/${UID2}/dir/f2
truncate -s 3 prefix/${UID3}/f1
truncate -s 3 prefix/${UID3}/f2
truncate -s 3 prefix/${UID3}/f3
ln -s prefix/${UID1}/f1 prefix/${UID1}/l1
ln -s prefix/${UID2}/f2 prefix/${UID2}/l1
ln -s prefix/${UID2}/f2 prefix/${UID2}/l2
ln -s prefix/${UID2}/f2 prefix/${UID2}/dir/l1
ln -s prefix/${UID2}/f2 prefix/${UID2}/dir/l2
ln -s prefix/${UID3}/f1 prefix/${UID3}/l1
ln -s prefix/${UID3}/f2 prefix/${UID3}/l2
ln -s prefix/${UID3}/f3 prefix/${UID3}/l3
sudo chown -R ${UID1}:${GID1} prefix/${UID1}
sudo chown -R ${UID2}:${GID2} prefix/${UID2}
sudo chown -R ${UID3}:${GID3} prefix/${UID3}
sudo chmod 700 prefix/${UID1}
sudo chmod 700 prefix/${UID2}
sudo chmod 700 prefix/${UID3}
sudo chmod 755 prefix
sudo chown 0:0 prefix

sudo ${ROOT}/src/gufi_dir2index -x prefix prefix.gufi

# replace text to get UIDs/GIDs 1001, 1002, and 1003
function replace() {
    echo "$@" | sed "s/${GUFI_STATS//\//\\/}/gufi_stats/g; s/\<${UID1}\>/1001/g; s/${UID2}\>/1002/g; s/\<${UID3}\>/1003/g; s/\<$(id -un)\>/1001/g; s/\<$(id -gn)\>/1001/g"
}

function root() {
    replace "# $@"
    output=$(sudo -E env "PATH=$PATH" env "PYTHONPATH=$PYTHONPATH" bash -c "$@")
    replace "${output}"
    echo
}

function user() {
    replace "$ $@"
    output=$(sudo -u \#${UID1} -E env "PATH=$PATH" env "PYTHONPATH=$PYTHONPATH" bash -c "$@")
    output="$($@)"
    replace "${output}"
    echo
}

(
root "${GUFI_STATS} depth"
root "${GUFI_STATS} depth -r"
root "${GUFI_STATS} depth ${UID2}"
root "${GUFI_STATS} depth ${UID2}/dir"
root "${GUFI_STATS} -r depth ${UID2}"
user "${GUFI_STATS} -r depth"

root "${GUFI_STATS} filesize"
root "${GUFI_STATS} -r filesize"
root "${GUFI_STATS} filesize ${UID2}"
root "${GUFI_STATS} -r filesize ${UID2}"
user "${GUFI_STATS} -r filesize"

root "${GUFI_STATS} filecount"
root "${GUFI_STATS} -r filecount"
root "${GUFI_STATS} filecount ${UID2}"
root "${GUFI_STATS} -r filecount ${UID2}"
user "${GUFI_STATS} -r filecount"

root "${GUFI_STATS} linkcount"
root "${GUFI_STATS} -r linkcount"
root "${GUFI_STATS} linkcount ${UID2}"
root "${GUFI_STATS} -r linkcount ${UID2}"
user "${GUFI_STATS} -r linkcount"

root "${GUFI_STATS} total-filesize"
root "${GUFI_STATS} -c total-filesize"
root "${GUFI_STATS} total-filesize ${UID2}"
root "${GUFI_STATS} -c total-filesize ${UID2}"
user "${GUFI_STATS} -c total-filesize"

root "${GUFI_STATS} total-filecount"
root "${GUFI_STATS} -c total-filecount"
root "${GUFI_STATS} total-filecount ${UID2}"
root "${GUFI_STATS} -c total-filecount ${UID2}"
user "${GUFI_STATS} -c total-filecount"

root "${GUFI_STATS} total-linkcount"
root "${GUFI_STATS} -c total-linkcount"
root "${GUFI_STATS} total-linkcount ${UID2}"
root "${GUFI_STATS} -c total-linkcount ${UID2}"
user "${GUFI_STATS} -c total-linkcount"

root "${GUFI_STATS} dircount"
root "${GUFI_STATS} -r dircount"
root "${GUFI_STATS} dircount ${UID2}"
root "${GUFI_STATS} -r dircount ${UID2}"
echo "###########################################"
echo "# This is wrong. This should be 3 but     #"
echo "# without root, the other databases can't #"
echo "# be accessed.                            #"
echo "###########################################"
user "${GUFI_STATS} -r dircount"

root "${GUFI_STATS} total-dircount"
root "${GUFI_STATS} -c total-dircount"
root "${GUFI_STATS} total-dircount ${UID2}"
root "${GUFI_STATS} -c total-dircount ${UID2}"
user "${GUFI_STATS} -c total-dircount"

root "${GUFI_STATS} leaf-dirs"
root "${GUFI_STATS} -r leaf-dirs"
root "${GUFI_STATS} leaf-dirs ${UID2}"
root "${GUFI_STATS} -r leaf-dirs ${UID2}"
user "${GUFI_STATS} -r leaf-dirs"

root "${GUFI_STATS} leaf-depth"
root "${GUFI_STATS} -r leaf-depth"
root "${GUFI_STATS} leaf-depth ${UID2}"
root "${GUFI_STATS} -r leaf-depth ${UID2}"
user "${GUFI_STATS} -r leaf-depth"

root "${GUFI_STATS} leaf-files"
root "${GUFI_STATS} -r leaf-files"
root "${GUFI_STATS} leaf-files ${UID2}"
root "${GUFI_STATS} -r leaf-files ${UID2}"
user "${GUFI_STATS} -r leaf-files"

root "${GUFI_STATS} total-leaf-files"
root "${GUFI_STATS} -c total-leaf-files"
root "${GUFI_STATS} total-leaf-files ${UID2}"
root "${GUFI_STATS} -c total-leaf-files ${UID2}"
user "${GUFI_STATS} -c total-leaf-files"

root "${GUFI_STATS} median-leaf-files"
root "${GUFI_STATS} median-leaf-files ${UID2}"
user "${GUFI_STATS} median-leaf-files"

root "${GUFI_STATS} files-per-level"
root "${GUFI_STATS} -c files-per-level"
root "${GUFI_STATS} files-per-level ${UID2}"
root "${GUFI_STATS} -c files-per-level ${UID2}"
user "${GUFI_STATS} -c files-per-level"

root "${GUFI_STATS} links-per-level"
root "${GUFI_STATS} -c links-per-level"
root "${GUFI_STATS} links-per-level ${UID2}"
root "${GUFI_STATS} -c links-per-level ${UID2}"
user "${GUFI_STATS} -c links-per-level"

root "${GUFI_STATS} dirs-per-level"
root "${GUFI_STATS} -c dirs-per-level"
root "${GUFI_STATS} dirs-per-level ${UID2}"
root "${GUFI_STATS} -c dirs-per-level ${UID2}"
user "${GUFI_STATS} -c dirs-per-level"
) | tee "${OUTPUT}"

diff -b ${ROOT}/test/gufi_stats.expected "${OUTPUT}"
