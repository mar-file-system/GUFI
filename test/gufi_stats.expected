# gufi_stats depth
0

# gufi_stats depth -r
. 0
1001 1
1002 1
1002/dir 2
1003 1

# gufi_stats depth 1002
1

# gufi_stats depth 1002/dir
2

# gufi_stats -r depth 1002
1002 1
1002/dir 2

$ gufi_stats -r depth
. 0
1001 1

# gufi_stats filesize
0

# gufi_stats -r filesize
. 0
1001 1
1002 4
1002/dir 4
1003 9

# gufi_stats filesize 1002
4

# gufi_stats -r filesize 1002
1002 4
1002/dir 4

$ gufi_stats -r filesize
. 0
1001 1

# gufi_stats filecount
0

# gufi_stats -r filecount
. 0
1001 1
1002 2
1002/dir 2
1003 3

# gufi_stats filecount 1002
2

# gufi_stats -r filecount 1002
1002 2
1002/dir 2

$ gufi_stats -r filecount
. 0
1001 1

# gufi_stats linkcount
0

# gufi_stats -r linkcount
. 0
1001 1
1002 2
1002/dir 2
1003 3

# gufi_stats linkcount 1002
2

# gufi_stats -r linkcount 1002
1002 2
1002/dir 2

$ gufi_stats -r linkcount
. 0
1001 1

# gufi_stats total-filesize
1001 1
1002 8
1003 9

# gufi_stats -c total-filesize
18

# gufi_stats total-filesize 1002
1002 8

# gufi_stats -c total-filesize 1002
8

$ gufi_stats -c total-filesize
1

# gufi_stats total-filecount
1001 1
1003 3
1002 4

# gufi_stats -c total-filecount
8

# gufi_stats total-filecount 1002
1002 4

# gufi_stats -c total-filecount 1002
4

$ gufi_stats -c total-filecount
1

# gufi_stats total-linkcount
1001 1
1003 3
1002 4

# gufi_stats -c total-linkcount
8

# gufi_stats total-linkcount 1002
1002 4

# gufi_stats -c total-linkcount 1002
4

$ gufi_stats -c total-linkcount
1

# gufi_stats dircount
3

# gufi_stats -r dircount
. 3
1001 0
1002 1
1002/dir 0
1003 0

# gufi_stats dircount 1002
1

# gufi_stats -r dircount 1002
1002 1
1002/dir 0

###########################################
# This is wrong. This should be 3 but     #
# without root, the other databases can't #
# be accessed.                            #
###########################################
$ gufi_stats -r dircount
. 1
1001 0

# gufi_stats total-dircount
1001 1
1003 1
1002 2

# gufi_stats -c total-dircount
4

# gufi_stats total-dircount 1002
1002 1

# gufi_stats -c total-dircount 1002
1

$ gufi_stats -c total-dircount
1

# gufi_stats leaf-dirs
1001
1003

# gufi_stats -r leaf-dirs
1001
1002/dir
1003

# gufi_stats leaf-dirs 1002
1002/dir

# gufi_stats -r leaf-dirs 1002
1002/dir

$ gufi_stats -r leaf-dirs
1001

# gufi_stats leaf-depth
1001 1
1003 1

# gufi_stats -r leaf-depth
1001 1
1002/dir 2
1003 1

# gufi_stats leaf-depth 1002
1002/dir 2

# gufi_stats -r leaf-depth 1002
1002/dir 2

$ gufi_stats -r leaf-depth
1001 1

# gufi_stats leaf-files
1001 1
1003 3

# gufi_stats -r leaf-files
1001 1
1002/dir 2
1003 3

# gufi_stats leaf-files 1002
1002/dir 2

# gufi_stats -r leaf-files 1002
1002/dir 2

$ gufi_stats -r leaf-files
1001 1

# gufi_stats total-leaf-files
1001 1
1002 2
1003 3

# gufi_stats -c total-leaf-files
6

# gufi_stats total-leaf-files 1002
1002 2

# gufi_stats -c total-leaf-files 1002
2

$ gufi_stats -c total-leaf-files
1

# gufi_stats median-leaf-files
2.0

# gufi_stats median-leaf-files 1002
2.0

$ gufi_stats median-leaf-files
1.0

# gufi_stats files-per-level
1001 1 1
1002 1 2
1003 1 3
1002 2 2

# gufi_stats -c files-per-level
1 6
2 2

# gufi_stats files-per-level 1002
1002 1 2
1002 2 2

# gufi_stats -c files-per-level 1002
1 2
2 2

$ gufi_stats -c files-per-level
1 1

# gufi_stats links-per-level
1001 1 1
1002 1 2
1003 1 3
1002 2 2

# gufi_stats -c links-per-level
1 6
2 2

# gufi_stats links-per-level 1002
1002 1 2
1002 2 2

# gufi_stats -c links-per-level 1002
1 2
2 2

$ gufi_stats -c links-per-level
1 1

# gufi_stats dirs-per-level
1001 1 1
1002 1 1
1003 1 1
1002 2 1

# gufi_stats -c dirs-per-level
1 3
2 1

# gufi_stats dirs-per-level 1002
1002 1 1
1002 2 1

# gufi_stats -c dirs-per-level 1002
1 1
2 1

$ gufi_stats -c dirs-per-level
1 1

